core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala

```diff
@@ -53,14 +53,22 @@ import org.apache.spark.util.{NextIterator, SerializableConfiguration, ShutdownH
 /**
  * A Spark split class that wraps around a Hadoop InputSplit.
  */
-private[spark] class HadoopPartition(rddId: Int, idx: Int, s: InputSplit)
+private[spark] class HadoopPartition(rddId: Int, override val index: Int, s: InputSplit)
   extends Partition {
 
   val inputSplit = new SerializableWritable[InputSplit](s)
 
-  override def hashCode(): Int = 41 * (41 + rddId) + idx
+  override def hashCode(): Int = 41 * (41 + rddId) + index
 
-  override val index: Int = idx
+  def canEqual(other: Any): Boolean = other.isInstanceOf[HadoopPartition]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: HadoopPartition =>
+      super.equals(that) &&
+        (that canEqual this) &&
+        index == that.index
```

Comment:
Lot's of questions... Do we need more people to review this? Or I can leave this for now but that will just report the issue to later.

Here is something I was thinking of:
```scala
private[spark] class HadoopPartition(
    private val rddId: Int,
    override val index: Int,
    s: InputSplit)
  extends Partition {

  val inputSplit = new SerializableWritable[InputSplit](s)

  override def hashCode(): Int = 41 * (41 + rddId) + index
  
  override def equals(other: Any): Boolean = other match {
    case that: HadoopPartition =>
      super.equals(that) &&
        rddId == that.rddId &&
        index == that.index
    case _ => false
  }
  ...
```